packages/network-controller/src/NetworkController.ts

```diff
@@ -92,6 +99,74 @@ export type NetworkConfiguration = {
   };
 };
 
+/**
+ * The collection of network configurations in state.
+ */
+type NetworkConfigurations = Record<
+  NetworkConfigurationId,
+  NetworkConfiguration & { id: NetworkConfigurationId }
+>;
+
+/**
+ * `Object.keys()` is intentionally generic: it returns the keys of an object,
+ * but it cannot make guarantees about the contents of that object, so the type
+ * of the keys is merely `string[]`. While this is technically accurate, it is
+ * also unnecessary if we have an object that we own and whose contents are
+ * known exactly.
+ *
+ * TODO: Move to @metamask/utils.
+ *
+ * @param object - The object.
+ * @returns The keys of an object, typed according to the type of the object
+ * itself.
+ */
+function knownKeysOf<K extends PropertyKey>(object: Partial<Record<K, any>>) {
+  return Object.keys(object) as K[];
+}
+
+/**
+ * Asserts that the given value is of the given type if the given functions
+ * returns a truthy result.
+ *
+ * @param _value - Some value.
+ * @param test - The function to run.
+ * @param message - The message to throw if the function does not return a
+ * truthy result.
+ * @throws if the function does not return a truthy result.
+ */
+function assertOfType<Type>(
+  _value: unknown,
+  test: () => boolean,
```

Comment:
Interesting idea, this is like a type guard builder.

Nit: Probably worth clarifying that `test` is meant to do the actual work of validation here? Maybe we could update the name and description to make that more clear, e.g. `validate - A function used to validate that the value is of the given type`.

Also, perhaps we could pass the value into this function? I know if `test` is declared inline it doesn't really make it easier to use that way, but if we're reusing validation functions from elsewhere that might be easier to work with.

Reply:
Sounds reasonable. Changes made here: [72d6d6e](https://github.com/MetaMask/core/pull/1439/commits/72d6d6e6392e62056d4fcef19e4a5d8135827327)